File: internal/sprites/sprites.go
// Package sprites scans a critterforge-generated critters/ directory and
// describes the sprite sheets it finds. It is shared by the kubagachi web
// server and the critterview development gallery.
package sprites

import (
	"encoding/binary"
	"fmt"
	"image"
	_ "image/png"
	"io"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"sync"
)

// States is the canonical state order used everywhere in the viewers.
// "failed" is accepted as an on-disk alias for "error".
var States = []string{
	"running",
	"pending",
	"completed",
	"crashloop",
	"backoff",
	"terminating",
	"unknown",
	"error",
}

var stateAliases = map[string][]string{
	"running":     {"running"},
	"pending":     {"pending"},
	"completed":   {"completed"},
	"crashloop":   {"crashloop"},
	"backoff":     {"backoff"},
	"terminating": {"terminating"},
	"unknown":     {"unknown"},
	"error":       {"error", "failed"},
}

// Info describes one critter's renderable artifacts.
type Info struct {
	Name          string             `json:"name"`
	KeyedURL      string             `json:"keyed_url,omitempty"`
	KeyedDim      *Dim               `json:"keyed_dim,omitempty"`
	KeyedHasAlpha bool               `json:"keyed_has_alpha"`
	Anim          map[string]AnimSrc `json:"anim,omitempty"` // state -> animation source
}

// Dim is a width/height pair.
type Dim struct {
	W int `json:"w"`
	H int `json:"h"`
}

// AnimSrc describes one per-state animation sheet.
type AnimSrc struct {
	URL      string        `json:"url"`
	W        int           `json:"w"`
	H        int           `json:"h"`
	Frames   int           `json:"frames"`
	HasAlpha bool          `json:"has_alpha"`
	Bounds   []FrameBounds `json:"bounds,omitempty"`
}

// FrameBounds is one detected frame's rectangle in the source sheet.
type FrameBounds struct {
	X0 int `json:"x0"`
	Y0 int `json:"y0"`
	X1 int `json:"x1"`
	Y1 int `json:"y1"`
}

// Scan walks dir and returns one Info per critter that has at least one
// renderable artifact. URLs are rooted at /critters/.
func Scan(dir string) ([]Info, error) {
	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil, err
	}
	out := []Info{}
	for _, e := range entries {
		if !e.IsDir() {
			continue
		}
		name := e.Name()
		info := Info{
			Name: name,
			Anim: map[string]AnimSrc{},
		}
		critterDir := filepath.Join(dir, name)

		keyedPath := filepath.Join(critterDir, "sprite-sheet-keyed.png")
		if d, hasAlpha, ok := pngInfo(keyedPath); ok {
			info.KeyedURL = "/critters/" + name + "/sprite-sheet-keyed.png?v=" + mtimeTag(keyedPath)
			info.KeyedDim = &d
			info.KeyedHasAlpha = hasAlpha
		}

		for _, state := range States {
			for _, alias := range stateAliases[state] {
				p := filepath.Join(critterDir, "sprite-sheet-"+alias+".png")
				if d, hasAlpha, ok := pngInfo(p); ok {
					info.Anim[state] = AnimSrc{
						URL:      "/critters/" + name + "/sprite-sheet-" + alias + ".png?v=" + mtimeTag(p),
						W:        d.W,
						H:        d.H,
						Frames:   8,
						HasAlpha: hasAlpha,
						Bounds:   DetectFrameBounds(p, 8),
					}
					break
				}
			}
		}

		// Workload animation decks (bursting, scaling, …) live alongside the base
		// states but aren't in the canonical States list. Discover any extra
		// sprite-sheet-<name>.png and serve it under its own key so the UI can
		// play it when a pod's critterState names it.
		if extra, err := filepath.Glob(filepath.Join(critterDir, "sprite-sheet-*.png")); err == nil {
			for _, p := range extra {
				base := filepath.Base(p)
				if base == "sprite-sheet-keyed.png" {
					continue
				}
				stem := strings.TrimSuffix(strings.TrimPrefix(base, "sprite-sheet-"), ".png")
				if _, exists := info.Anim[stem]; exists {
					continue
				}
				if d, hasAlpha, ok := pngInfo(p); ok {
					info.Anim[stem] = AnimSrc{
						URL:      "/critters/" + name + "/" + base + "?v=" + mtimeTag(p),
						W:        d.W,
						H:        d.H,
						Frames:   8,
						HasAlpha: hasAlpha,
						Bounds:   DetectFrameBounds(p, 8),
					}
				}
			}
		}

		if info.KeyedURL == "" && len(info.Anim) == 0 {
			continue
		}
		out = append(out, info)
	}
	sort.Slice(out, func(i, j int) bool { return strings.ToLower(out[i].Name) < strings.ToLower(out[j].Name) })
	return out, nil
}

// LatestItem describes one PNG for the flat "most recent artifacts" list.
type LatestItem struct {
	Critter  string        `json:"critter"`
	State    string        `json:"state,omitempty"`
	Kind     string        `json:"kind"` // "anim" | "keyed" | "tile"
	URL      string        `json:"url"`
	Filename string        `json:"filename"`
	W        int           `json:"w"`
	H        int           `json:"h"`
	Frames   int           `json:"frames,omitempty"`
	HasAlpha bool          `json:"has_alpha"`
	Bounds   []FrameBounds `json:"bounds,omitempty"`
	ModUnix  int64         `json:"mod_unix"`
	ModISO   string        `json:"mod_iso"`
}

// ScanLatest walks the critters tree, classifies every PNG, and returns the
// limit most recently modified files.
func ScanLatest(dir string, limit int) ([]LatestItem, error) {
	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil, err
	}
	out := []LatestItem{}
	for _, e := range entries {
		if !e.IsDir() {
			continue
		}
		name := e.Name()
		critterDir := filepath.Join(dir, name)
		files, err := os.ReadDir(critterDir)
		if err != nil {
			continue
		}
		for _, f := range files {
			if f.IsDir() || !strings.HasSuffix(f.Name(), ".png") {
				continue
			}
			full := filepath.Join(critterDir, f.Name())
			info, err := os.Stat(full)
			if err != nil {
				continue
			}
			d, hasAlpha, ok := pngInfo(full)
			if !ok {
				continue
			}
			kind, state, frames := classifyFilename(f.Name())
			var bounds []FrameBounds
			if kind == "anim" && frames > 0 {
				bounds = DetectFrameBounds(full, frames)
			}
			out = append(out, LatestItem{
				Critter:  name,
				State:    state,
				Kind:     kind,
				URL:      "/critters/" + name + "/" + f.Name() + "?v=" + mtimeTag(full),
				Filename: f.Name(),
				W:        d.W,
				H:        d.H,
				Frames:   frames,
				HasAlpha: hasAlpha,
				Bounds:   bounds,
				ModUnix:  info.ModTime().Unix(),
				ModISO:   info.ModTime().UTC().Format("2006-01-02T15:04:05Z"),
			})
		}
	}
	sort.Slice(out, func(i, j int) bool { return out[i].ModUnix > out[j].ModUnix })
	if len(out) > limit {
		out = out[:limit]
	}
	return out, nil
}

func classifyFilename(name string) (kind, state string, frames int) {
	if name == "sprite-sheet-keyed.png" || name == "sprite-sheet.png" {
		return "keyed", "", 0
	}
	if strings.HasPrefix(name, "sprite-sheet-") && strings.HasSuffix(name, ".png") {
		stem := strings.TrimSuffix(strings.TrimPrefix(name, "sprite-sheet-"), ".png")
		if stem == "failed" {
			stem = "error"
		}
		return "anim", stem, 8
	}
	stem := strings.TrimSuffix(name, ".png")
	if stem == "failed" {
		stem = "error"
	}
	return "tile", stem, 0
}

// mtimeTag returns a short cache-busting token derived from the file's
// modification time.
func mtimeTag(path string) string {
	info, err := os.Stat(path)
	if err != nil {
		return "0"
	}
	return fmt.Sprintf("%d", info.ModTime().UnixNano())
}

// boundsCache memoizes column-scan frame detection by (path, mtimeNano).
var (
	boundsCacheMu sync.RWMutex
	boundsCache   = map[string][]FrameBounds{}
)

// DetectFrameBounds returns one rect per detected frame in the PNG at path.
// Caller passes the expected frame count for fallback / sanity checking.
func DetectFrameBounds(path string, expected int) []FrameBounds {
	info, err := os.Stat(path)
	if err != nil {
		return nil
	}
	key := fmt.Sprintf("%s:%d", path, info.ModTime().UnixNano())

	boundsCacheMu.RLock()
	if v, ok := boundsCache[key]; ok {
		boundsCacheMu.RUnlock()
		return v
	}
	boundsCacheMu.RUnlock()

	v := computeFrameBounds(path, expected)
	if v != nil {
		boundsCacheMu.Lock()
		boundsCache[key] = v
		boundsCacheMu.Unlock()
	}
	return v
}

func computeFrameBounds(path string, expected int) []FrameBounds {
	f, err := os.Open(path)
	if err != nil {
		return nil
	}
	defer f.Close()
	src, _, err := image.Decode(f)
	if err != nil {
		return nil
	}
	b := src.Bounds()
	w := b.Dx()
	h := b.Dy()
	if w == 0 || h == 0 {
		return nil
	}

	// Per-column accounting: a column needs enough opaque pixels to count as
	// "inside a frame" so stray glitter doesn't bridge two real frames.
	const alphaMin = 24
	colCount := make([]int, w)
	colY0 := make([]int, w)
	colY1 := make([]int, w)
	for i := range colY0 {
		colY0[i] = h
		colY1[i] = -1
	}

	mark := func(x, y int) {
		colCount[x]++
		if y < colY0[x] {
			colY0[x] = y
		}
		if y > colY1[x] {
			colY1[x] = y
		}
	}

	switch img := src.(type) {
	case *image.NRGBA:
		for y := 0; y < h; y++ {
			row := y * img.Stride
			for x := 0; x < w; x++ {
				if img.Pix[row+x*4+3] > alphaMin {
					mark(x, y)
				}
			}
		}
	case *image.RGBA:
		for y := 0; y < h; y++ {
			row := y * img.Stride
			for x := 0; x < w; x++ {
				if img.Pix[row+x*4+3] > alphaMin {
					mark(x, y)
				}
			}
		}
	default:
		for y := 0; y < h; y++ {
			for x := 0; x < w; x++ {
				_, _, _, a := src.At(b.Min.X+x, b.Min.Y+y).RGBA()
				if a>>8 > alphaMin {
					mark(x, y)
				}
			}
		}
	}

	densityMin := h / 25
	if densityMin < 6 {
		densityMin = 6
	}
	colHasOpaque := make([]bool, w)
	for x := 0; x < w; x++ {
		if colCount[x] >= densityMin {
			colHasOpaque[x] = true
		}
	}

	gapTolerance := w / (expected * 12)
	if gapTolerance < 2 {
		gapTolerance = 2
	}
	var clusters []FrameBounds
	var cur *FrameBounds
	gapRun := 0
	for x := 0; x < w; x++ {
		if colHasOpaque[x] {
			if cur == nil {
				cur = &FrameBounds{X0: x, X1: x, Y0: colY0[x], Y1: colY1[x]}
			} else {
				cur.X1 = x
				if colY0[x] < cur.Y0 {
					cur.Y0 = colY0[x]
				}
				if colY1[x] > cur.Y1 {
					cur.Y1 = colY1[x]
				}
			}
			gapRun = 0
		} else if cur != nil {
			gapRun++
			if gapRun > gapTolerance {
				cur.X1++ // exclusive
				cur.Y1++
				clusters = append(clusters, *cur)
				cur = nil
				gapRun = 0
			}
		}
	}
	if cur != nil {
		cur.X1++
		cur.Y1++
		clusters = append(clusters, *cur)
	}
	if len(clusters) == 0 {
		return nil
	}
	return clusters
}

// pngInfo returns the PNG's width/height plus whether it carries an alpha
// channel, by reading the IHDR chunk directly.
func pngInfo(path string) (Dim, bool, bool) {
	f, err := os.Open(path)
	if err != nil {
		return Dim{}, false, false
	}
	defer f.Close()

	var head [26]byte
	if _, err := io.ReadFull(f, head[:]); err != nil {
		return Dim{}, false, false
	}
	if string(head[0:8]) != "\x89PNG\r\n\x1a\n" {
		return Dim{}, false, false
	}
	if string(head[12:16]) != "IHDR" {
		return Dim{}, false, false
	}
	w := int(binary.BigEndian.Uint32(head[16:20]))
	h := int(binary.BigEndian.Uint32(head[20:24]))
	colorType := head[25]
	hasAlpha := colorType == 4 || colorType == 6
	return Dim{W: w, H: h}, hasAlpha, true
}

File: native/src/sprites.rs
//! Sprite sheet discovery for /api/critters, ported from internal/sprites.
//! Scan walks the critters dir; per critter it reads the keyed sheet header
//! and each per-state animation deck, and detects per-frame content bounds
//! (equal-width frame split, tight opaque bbox per frame).

use std::collections::{BTreeMap, HashMap};
use std::io::Read as _;
use std::path::Path;
use std::sync::Mutex;

use serde::Serialize;

/// Canonical state order used everywhere in the viewers. "failed" is accepted
/// as an on-disk alias for "error".
pub const STATES: [&str; 8] = [
    "running",
    "pending",
    "completed",
    "crashloop",
    "backoff",
    "terminating",
    "unknown",
    "error",
];

fn state_aliases(state: &str) -> &'static [&'static str] {
    match state {
        "running" => &["running"],
        "pending" => &["pending"],
        "completed" => &["completed"],
        "crashloop" => &["crashloop"],
        "backoff" => &["backoff"],
        "terminating" => &["terminating"],
        "unknown" => &["unknown"],
        "error" => &["error", "failed"],
        _ => &[],
    }
}

#[derive(Serialize, Clone, Debug)]
pub struct Dim {
    pub w: u32,
    pub h: u32,
}

#[derive(Serialize, Clone, Debug)]
pub struct FrameBounds {
    pub x0: u32,
    pub y0: u32,
    pub x1: u32,
    pub y1: u32,
}

#[derive(Serialize, Clone, Debug)]
pub struct AnimSrc {
    pub url: String,
    pub w: u32,
    pub h: u32,
    pub frames: u32,
    pub has_alpha: bool,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub bounds: Vec<FrameBounds>,
}

#[derive(Serialize, Clone, Debug)]
pub struct Info {
    pub name: String,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub keyed_url: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub keyed_dim: Option<Dim>,
    pub keyed_has_alpha: bool,
    #[serde(skip_serializing_if = "BTreeMap::is_empty")]
    pub anim: BTreeMap<String, AnimSrc>,
}

/// Memoizes frame detection by (path, mtimeNanos), like the Go boundsCache.
pub type BoundsCache = Mutex<HashMap<String, Vec<FrameBounds>>>;

fn mtime_nanos(path: &Path) -> u128 {
    std::fs::metadata(path)
        .and_then(|m| m.modified())
        .ok()
        .and_then(|t| t.duration_since(std::time::UNIX_EPOCH).ok())
        .map(|d| d.as_nanos())
        .unwrap_or(0)
}

/// png_info returns the PNG's width/height plus whether it carries an alpha
/// channel, by reading the IHDR chunk directly.
fn png_info(path: &Path) -> Option<(Dim, bool)> {
    let mut f = std::fs::File::open(path).ok()?;
    let mut head = [0u8; 26];
    f.read_exact(&mut head).ok()?;
    if &head[0..8] != b"\x89PNG\r\n\x1a\n" {
        return None;
    }
    if &head[12..16] != b"IHDR" {
        return None;
    }
    let w = u32::from_be_bytes([head[16], head[17], head[18], head[19]]);
    let h = u32::from_be_bytes([head[20], head[21], head[22], head[23]]);
    let color_type = head[25];
    let has_alpha = color_type == 4 || color_type == 6;
    Some((Dim { w, h }, has_alpha))
}

/// detect_frame_bounds returns one rect per frame with visible content: the
/// sheet is split into `expected` equal-width frames and each frame's tight
/// opaque bbox (alpha > 0) is reported with exclusive x1/y1. Fully transparent
/// frames are skipped.
pub fn detect_frame_bounds(path: &Path, expected: u32, cache: &BoundsCache) -> Vec<FrameBounds> {
    let key = format!("{}:{}", path.display(), mtime_nanos(path));
    if let Some(v) = cache.lock().unwrap().get(&key) {
        return v.clone();
    }
    let v = compute_frame_bounds(path, expected);
    if !v.is_empty() {
        cache.lock().unwrap().insert(key, v.clone());
    }
    v
}

fn compute_frame_bounds(path: &Path, expected: u32) -> Vec<FrameBounds> {
    let img = match image::open(path) {
        Ok(i) => i,
        Err(_) => return Vec::new(),
    };
    let rgba = img.to_rgba8();
    let (w, h) = rgba.dimensions();
    if w == 0 || h == 0 || expected == 0 {
        return Vec::new();
    }
    let buf = rgba.as_raw();
    let stride = (w * 4) as usize;

    let mut out = Vec::new();
    for i in 0..expected {
        let fx0 = i * w / expected;
        let fx1 = (i + 1) * w / expected;
        let (mut min_x, mut min_y) = (u32::MAX, u32::MAX);
        let (mut max_x, mut max_y) = (0u32, 0u32);
        let mut any = false;
        for y in 0..h {
            let row = y as usize * stride;
            for x in fx0..fx1 {
                if buf[row + x as usize * 4 + 3] > 0 {
                    any = true;
                    if x < min_x {
                        min_x = x;
                    }
                    if x > max_x {
                        max_x = x;
                    }
                    if y < min_y {
                        min_y = y;
                    }
                    if y > max_y {
                        max_y = y;
                    }
                }
            }
        }
        if any {
            out.push(FrameBounds {
                x0: min_x,
                y0: min_y,
                x1: max_x + 1, // exclusive
                y1: max_y + 1,
            });
        }
    }
    out
}

/// scan walks dir and returns one Info per critter that has at least one
/// renderable artifact. URLs are rooted at /critters/.
pub fn scan(dir: &Path, cache: &BoundsCache) -> Vec<Info> {
    let mut out = Vec::new();
    let entries = match std::fs::read_dir(dir) {
        Ok(e) => e,
        Err(_) => return out,
    };
    for e in entries.flatten() {
        if !e.file_type().map(|t| t.is_dir()).unwrap_or(false) {
            continue;
        }
        let name = e.file_name().to_string_lossy().to_string();
        let critter_dir = dir.join(&name);
        let mut info = Info {
            name: name.clone(),
            keyed_url: String::new(),
            keyed_dim: None,
            keyed_has_alpha: false,
            anim: BTreeMap::new(),
        };

        let keyed = critter_dir.join("sprite-sheet-keyed.png");
        if let Some((d, has_alpha)) = png_info(&keyed) {
            info.keyed_url = format!(
                "/critters/{}/sprite-sheet-keyed.png?v={}",
                name,
                mtime_nanos(&keyed)
            );
            info.keyed_dim = Some(d);
            info.keyed_has_alpha = has_alpha;
        }

        for state in STATES {
            for alias in state_aliases(state) {
                let p = critter_dir.join(format!("sprite-sheet-{alias}.png"));
                if let Some((d, has_alpha)) = png_info(&p) {
                    info.anim.insert(
                        state.to_string(),
                        AnimSrc {
                            url: format!(
                                "/critters/{}/sprite-sheet-{}.png?v={}",
                                name,
                                alias,
                                mtime_nanos(&p)
                            ),
                            w: d.w,
                            h: d.h,
                            frames: 8,
                            has_alpha,
                            bounds: detect_frame_bounds(&p, 8, cache),
                        },
                    );
                    break;
                }
            }
        }

        // Workload animation decks (bursting, scaling, …) live alongside the
        // base states but aren't in the canonical list. Discover any extra
        // sprite-sheet-<name>.png and serve it under its own key.
        if let Ok(files) = std::fs::read_dir(&critter_dir) {
            let mut extra: Vec<String> = files
                .flatten()
                .filter_map(|f| {
                    let n = f.file_name().to_string_lossy().to_string();
                    (n.starts_with("sprite-sheet-") && n.ends_with(".png")).then_some(n)
                })
                .collect();
            extra.sort();
            for base in extra {
                if base == "sprite-sheet-keyed.png" {
                    continue;
                }
                let stem = base
                    .strip_prefix("sprite-sheet-")
                    .and_then(|s| s.strip_suffix(".png"))
                    .unwrap_or(&base)
                    .to_string();
                if info.anim.contains_key(&stem) {
                    continue;
                }
                let p = critter_dir.join(&base);
                if let Some((d, has_alpha)) = png_info(&p) {
                    info.anim.insert(
                        stem,
                        AnimSrc {
                            url: format!("/critters/{}/{}?v={}", name, base, mtime_nanos(&p)),
                            w: d.w,
                            h: d.h,
                            frames: 8,
                            has_alpha,
                            bounds: detect_frame_bounds(&p, 8, cache),
                        },
                    );
                }
            }
        }

        if info.keyed_url.is_empty() && info.anim.is_empty() {
            continue;
        }
        out.push(info);
    }
    out.sort_by(|a, b| a.name.to_lowercase().cmp(&b.name.to_lowercase()));
    out
}
